### ISSUE TITLE
ABC iView restricting number of shows in the feed

### ISSUE THREAD

**Issue by KODeKarnage** [andybotting/xbmc-addon-abc-iview#1928]
I've noticed that iView is providng only a small number of shows (1 or 2 in most cases) in the feed response. Any idea about what is causing this behaviour? Just a change in policy from ABC? I couldnt find documentation on their api.

Cheers,

**Comment by andybotting** [comment]
Documentation on their API? That's a good one :)

Generally the rule is they have programs for two weeks, but some other
programs might have a full season.

On 2 Sep 2016 11:01 PM, "KODeKarnage" <notifications@github.com> wrote:

I've noticed that iView is providng only a small number of shows (1 or 2 in
most cases) in the feed response. Any idea about what is causing this
behaviour? Just a change in policy from ABC? I couldnt find documentation
on their api.

Cheers,

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/andybotting/xbmc-addon-abc-iview/issues/1928>, or mute
the thread

**Comment by andybotting** [comment]
Maybe do a comparison against what is returned from the mobile app? We use
the smart TV API for our access

**Comment by glennguy** [comment]
Can see the same here, in particular the kids feed is a fraction of what it used to be whereas web/mobile is as normal.

Some recent posts on the Whirlpool iView thread are saying how their smart TV apps have stopped working altogether. I've just tried the PS4 app and that didn't work at all. My 2015 Samsung was fine although the functioning of the app is very similar to the mobile app so maybe they use the same API. 

It might be worth looking into switching to the mobile API, who knows if this is just a backend stuffup or a possible decommissioning of lesser platforms. I'll try and find some time this weekend to get a fork using the mobile API.

I've had a look into it and the biggest decision is deciding how to handle the list of episodes for a show - there's no way in the mobile API to query a complete list. The list of shows from each category will only contain a link to the most recent episode for each show, which in turn only contains full metadata for itself and only links to the other episodes. So would we be better to get full metadata including URL for only the first item in the list and calculate the others at playback time or do it all beforehand? On the plus side grabbing series lists for the larger categories and even the full a-z listing will take no time at all.

**Comment by rocket-pants** [comment]
Same issue. Only returning 2 play school episodes with 38 on website / android app (other shows are also affected but play school is the only one to cause tears).  Checking to see how the Xbox app fares.

**Comment by rocket-pants** [comment]
The Xbox app always to be working fine

**Comment by trideceth12** [comment]
Confirming the same for me. I get the full list of shows from other sources (including webdl) but not from your wonderful kodi plugin.

**Comment by cagney** [comment]
&lt;abc&gt;me too!&lt;/abc&gt;

**Comment by andybotting** [comment]
OK, looks like we have some work to do to switch over to the mobile app API!

**Comment by andybotting** [comment]
@rocket-pants do you get the full list from your Samsung TV? The APIs for each of the smart TV brands are pretty similar. We're using the LG feed currently, but could switch to Samsung without any real code change. I'll check my LG TV tonight

**Comment by rocket-pants** [comment]
I go t a mate to check, so I'll ask again but he said there were "heaps" on the Samsung.

**Comment by andybotting** [comment]
Yeah, I just looked at my LG TV and found that Sesame Street had a stack of episodes, where we only show 2. Will investigate.

**Comment by glennguy** [comment]
I'm half way there with the mobile API - I have categories and programmes worked out, will get episodes and playback done tonight or tomorrow night and report back!

**Comment by andybotting** [comment]
Nice work as always @glennguy 

I can see the issue isn't which API we're using, it's how we're requesting the program listing. When we fetch the category (e.g. ABC4Kids) we get a limited list of programs. If I search or go via the S-Z category, I see more. 

I have a feeling the mobile app might be the same. We'll need to restructure our addon to more closely follow the structure of iView to make better use of the API I think.

**Comment by andybotting** [comment]
Just had a look at the mobile app - definitely need to change. I'll look forward to seeing your work @glennguy

**Comment by cagney** [comment]
can "master.zip" please be tweaked so it is easier to tell that we're really running latest with these changes?

**Comment by andybotting** [comment]
I've pulled the changes into a new branch called mobile-api. Of you switch to that branch, you should be able to grab a zip file of that

**Comment by rocket-pants** [comment]
Tested the mobile API-branch and. while  I haven't done extensive testing, it appears to be returning the same list as the website version of iview. Win10 Kodi Isengard 15.2

Thank very much for the hard work on this.

Cheers

**Comment by andybotting** [comment]
@rocket-pants thanks for the feedback. I've seen issues where stream fail to load after some time, and only a restart of Kodi and/or manually deleting cookies.dat fixes it. Let me know if you see anything like that - or any other issues for that matter. Cheers

**Comment by rocket-pants** [comment]
will do Andy, Thanks again Andy and Glenn. You've made my daughter very
happy.  Any chance you can drop a note here once it is merged into master
so I know to go back to that version?

Ed

**Comment by glennguy** [comment]
You're welcome Ed - I have a young daughter too, you can guess where some
of my motivations came from!

**Comment by cagney** [comment]
wgot the zip file around midnight GMT and up popped all the shows and things worked, ya!  Unfortunately, now, they don't seem to want to play, but that could easily be one of the other ducks we chase here.  Will play some more.

Thanks.

(need a way to brand the zip)

**Comment by glennguy** [comment]
@cagney are you able to provide a debug log? Also try deleting your
cookies.dat from the cache folder.

**Comment by cagney** [comment]
One dead DSL connection later ....

Deleting cookies.dat got things working, ya!  I'll see what logs I can dig up.

**Comment by andybotting** [comment]
@cagney I think switching between live streams and the catchup programs is what causes the fail. Is that what you were seeing?

**Comment by cagney** [comment]
@andybotting yes, after trying a live stream things fail.  For instance:

15:43:19 T:2757002304  NOTICE: -->Python Interpreter Initialized<--
15:43:19 T:2757002304  NOTICE: [ABC iView v1.7.3] XBMC/Kodi 15.2 running on Linux
15:43:22 T:3041857536  NOTICE: DVDPlayer: Opening: http://iviewhls-i.akamaihd.net/i/playback/_definst_/_video/kids_angelinacgi_04_11_,650000,494000,205000,40748,.mp4.csmil/master.m3u8?h
dnea=st=1473968000~exp=1473975200~acl=/*~hmac=c1777d82ff93e329c5c357b9c53f7f2f6a4d6468babbb88552ab6f74d4ed1641
15:43:22 T:3041857536 WARNING: CDVDMessageQueue(player)::Put MSGQ_NOT_INITIALIZED
15:43:22 T:2757002304  NOTICE: Thread DVDPlayer start, auto delete: false
15:43:22 T:2757002304  NOTICE: Creating InputStream
15:43:23 T:2929718336  NOTICE: Thread FileCache start, auto delete: false
15:43:23 T:2757002304  NOTICE: Creating Demuxer
15:43:25 T:2757002304   ERROR: OpenDemuxStream - Error creating demuxer
15:43:25 T:2757002304  NOTICE: CDVDPlayer::OnExit()
15:43:25 T:3041857536   ERROR: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.abc_iview/?play=true&rating=G&date=2016-09-15+06%3A35%3A00&description=Angelina+
%26+her+classmates+are+given+the+assignment+to+create+a+nature-inspired+dance+routine.&title=Angelina+Ballerina%3A+The+Next+Steps&url=http%3A%2F%2Fiviewhls-i.akamaihd.net%2Fi%2Fplayback
%2F_definst_%2F_video%2Fkids_angelinacgi_04_11_%2C650000%2C494000%2C205000%2C40748%2C.mp4.csmil%2Fmaster.m3u8&house_number=ZX1217A011S00&thumbnail=http%3A%2F%2Fcdn.iview.abc.net.au%2Fth
umbs%2Fi%2Fzx%2FZX1217A011S0055bedbc0b50b21.04560999_1280.jpg&episode_title=Angelina%27s+Nature+Dance&duration=790&subtitle_url=http%3A%2F%2Fiview.abc.net.au%2Fcc%2Fzx%2FZX1217A011S00MA
1D1_57d7e4a4f0f14.xml&id=-1]

**Comment by andybotting** [comment]
@cagney thanks for confirmation

**Comment by glennguy** [comment]
@andybotting I've got a little bit more insight into what's happening with the cookies.

What I've found is if there's a 'hdntl' cookie in the cookies.dat that is newer (on a line lower) and for a different subdomain (or even whole domain) than for the stream we're trying to play, then it will fail. Where this will cause problems is for the plethora of other content providers that use akamaihd's services. Plus7 and SBS use them but don't rely on the hdntl cookie for security so they're not conflicting at least.
So if you close Kodi and add a new cookie in cookies.dat for any domain, as long as the cookie name is 'hdntl' it will stop playback for everything on iView.

In pseudo-code, it's like the logic that FFmpeg is reading the cookies.dat looks like:
```
for cookie in 'cookies.dat'.reversed()
    if cookie.name == 'hdntl'
        if cookie.domain == currentdomain:
            return cookie
        else:
            return none
```
Which stops looking after the first 'hdntl' cookie.

It would be solved if Kodi deleted session cookies after each playlist item has finished.

**Comment by mstrangi** [comment]
Hi Andy/Glenn,
Great work for supporting this and other catchup tv addons - it's very much appreciated. I am no expert developer but happy to help out with any testing on RPI3 or 2. My son will also appreciate this fix.

**Comment by cagney** [comment]
BTW, my workaround was to add the following:
```
OpenELEC:~ # cat /storage/.config/autostart.sh 
#!/bin/sh
rm -f /storage/.kodi/temp/cookies.dat
```
so a reboot flushes the problem :-/

**Comment by glennguy** [comment]
The cookie issue should be all sorted in #1958. I put in a bug report the other week for Kodi's side of things and it is fixed in beta 3 and newer - cookies.dat is no longer used.
